src/bvals/comms/tag_map.cpp

```diff
@@ -35,13 +35,34 @@ TagMap::rank_pair_t TagMap::MakeChannelPair(const MeshBlock *pmb,
 }
 template <BoundaryType BOUND>
 void TagMap::AddMeshDataToMap(std::shared_ptr<MeshData<Real>> &md) {
-  ForEachBoundary<BOUND>(md, [&](auto pmb, sp_mbd_t rc, nb_t &nb, const sp_cv_t v) {
-    const int other_rank = nb.snb.rank;
-    if (map_.count(other_rank) < 1) map_[other_rank] = rank_pair_map_t();
-    auto &pair_map = map_[other_rank];
-    // Add channel key with an invalid tag
-    pair_map[MakeChannelPair(pmb, nb)] = -1;
-  });
+  for (int block = 0; block < md->NumBlocks(); ++block) {
+    auto &rc = md->GetBlockData(block);
+    auto pmb = rc->GetBlockPointer();
+    auto *neighbors = [&pmb, &md] {
+      if constexpr (BOUND == BoundaryType::gmg_restrict_send)
+        return &(pmb->gmg_coarser_neighbors);
+      if constexpr (BOUND == BoundaryType::gmg_restrict_recv)
+        return &(pmb->gmg_finer_neighbors);
+      if constexpr (BOUND == BoundaryType::gmg_prolongate_send)
+        return &(pmb->gmg_finer_neighbors);
+      if constexpr (BOUND == BoundaryType::gmg_prolongate_recv)
+        return &(pmb->gmg_coarser_neighbors);
+      if constexpr (BOUND == BoundaryType::gmg_prolongate_recv)
+        return &(pmb->gmg_coarser_neighbors);
+      if constexpr (BOUND == BoundaryType::gmg_same)
+        return pmb->loc.level() == md->grid.logical_level
+                   ? &(pmb->gmg_same_neighbors)
+                   : &(pmb->gmg_composite_finer_neighbors);
+      return &(pmb->neighbors);
+    }();
```

Comment:
Yes, it creates a lambda and then calls it at that location. I tend to like that pattern more for more complex initializations. As you say, in this case I could do 
```c++
std::vector<NeighborBlock> *neighbors;
if constexpr(....) {
  neighbors = &(pmb->gmg_coarser_neighbors);
} else if (...) {
...
```
but I guess I prefer the pattern that is in there now.

Reply:
I find the lambda version confusing, but I won't fight you on it. I was able to disentangle what's going on.

Reply:
yeah, obviously it doesn't really buy you anything for pointer initialization but I think it is a useful pattern since default construction of other types of objects can be expensive. It is also a useful pattern for use in class initializer lists.

Reply:
Just to chime in, this syntax would also be confusing to me (since I can see this discussion I now know what's going on) but it might be worthwhile to just add a comment saying `define and use lambda to reduce initializations while setting neighbors` so people like me don't have to wonder whether there is something more or different going on.

Reply:
I would be happy with a comment to that effect.

Reply:
comment added.